### PR TITLE
Resolve duplicate error - backend (Brendan + Taya) + frontend (Taya)

### PR DESCRIPTION
**Pull request by tyakovenko**
issue#11
To run: 
- change the window size parameter in default-config.go if needed (right now it's set to 2 to make it easier to see if the system is working without having to wait long)
- in analysis_handler.go change maxPoints as needed to display a particular number of points on the graph 
- to see the full result struct output go to localhost:8080/data
- Note: both of those become config parameters and will be pulled into this branch once it combines with main/jolene's pr

major file changes:
-  refactored the result and outlier structs in common/structures.go to link everything by timestamps (helps in getting the duplicates)
- static/script.js refactored data retrival to be consistent with the new data types; added timestamps as labels to the x-axis
- static/index.html removed some object and only kept the graph and boxes for the outliers + their timestamps (the filtering; config parameters, and message displays are part of other issues and will be in a different pull request)
- analysis_handler.go refactored result recording to make the new struct; refactored algorithm calls (in blt_mad and shake_alert) to better fit the new data structs

### CODE REVIEW

File: src/parse/ris_live.go [hunk 1]
```diff
@@ -169,7 +169,8 @@ func receiveHandler(msgChannel chan common.Message, conn *websocket.Conn) {
 		if err != nil {
 			log.Println("Error parsing BGP message", err)
 		} else {
-			fmt.Printf("Parsed BGP Message: %+v\n", bgpMsgs) //prints parsed BGP msg
+			fmt.Println("")
```
Comment (sdstrowes) [inline src/parse/ris_live.go:172]: ?

File: src/default-config.json [hunk 2]
```diff
@@ -7,5 +7,5 @@
   "asn": "enter the ASNs of interest here",
   "peerIp": "enter peer IP",
   "connector": "enter connector or connectors of interests",
-  "windowSize": "360"
+  "windowSize": "2"
```
Comment (sdstrowes) [inline src/default-config.json:10]: assume this should be changed back to a higher value (but a windowSize of 2 or even 1 is a great test of boundary conditions in the windowing/anomaly detection code, even if not useful for real workloads)

File: src/blt_mad/helpers.go [hunk 3]
```diff
@@ -101,100 +97,3 @@ func WithinToleranceFloatSlice(a []float64, b []float64, e float64) bool {
 	}
 	return true
 }
-
-func calculatePercentile(numbers []float64, percentile float64) float64 {
```
Comment (sdstrowes) [inline src/blt_mad/helpers.go:105]: have these moved somewhere, or were they simply not used?

File: src/shake_alert/shake_alert.go [hunk 4]
```diff
@@ -60,16 +60,20 @@ func findR(data []float64) float64 {
...
 // input: array of ints, where each index represents a time bin and value is the count for that time bin
 // output: values of outliers aka counts
-func FindOutliers(data []float64) []float64 {
+func FindOutliers(data []float64) ([]float64, []int) {
 	//result of outliers
 	var outliers []float64
+	var indexes []int
+
+	dataCopy := make([]float64, len(data))
+	copy(dataCopy, data)
```
Comment (sdstrowes) [inline src/shake_alert/shake_alert.go:69]: What's the purpose in the `copy()` here?

Comment (bman4) [reply src/shake_alert/shake_alert.go:69]: There was a small bug where the findR function, would sort the actual array we were analyzing to calculate R. So we just gave it a copy instead.

File: src/analyze/analysis_handler.go [hunk 5]
```diff
@@ -6,85 +6,148 @@ import (
...
-	//put all the results into the Result struct and pass write it out to a json
-	r := common.Result{
-		Prefix:     config.Prefix,
-		ASN:        config.Asn,
-		PeerIP:     config.PeerIP,
-		WindowSize: config.WindowSize,
-
-		Frequencies: sortedFrequencies, //make all of these maps and append onto them here; create them as global vars
+	//check the amount of the frequencies in results
+	if (len(AllResults.AllFreq) + 1) > maxPoints { //adding one more frequency would result in more than needed points
```
Comment (apai-fsly) [inline src/analyze/analysis_handler.go:40]: on line 30 you defined the frequencies variable. I would suggest using that as the conditional.

File: src/analyze/analysis_handler.go [hunk 6]
```diff
@@ -6,85 +6,148 @@ import (
...
+		//remove the first item in the freq map; no modifications to the outlier map
+		firstResultKey := getSmallestTimestamp(getListOfKeys(frequencies))
+		delete(frequencies, firstResultKey)
+
+		//now update the results
+		for timestamp, _ := range lengthMap {
```
Comment (apai-fsly) [inline src/analyze/analysis_handler.go:46]: If you only need the keys for the map you can remove ", _"

Comment (salmans) [reply src/analyze/analysis_handler.go:46]: huh, I didn't know that. Is that a new feature in Go or has been around forever?

File: src/static/script.js [hunk 7]
```diff
@@ -9,49 +11,40 @@ function addData(chart, newData) {
...
+
+function addLabels(allOutliers) {
+    const outlierList = Object.values(allOutliers)
+
+    const timestampList = [];
+    const valsList = [];
+
+    for (let i = 0; i < outlierList.length; i++) {
+        timestamp = new Date(outlierList[i].Timestamp);
+        count =outlierList[i].Count;
```
Comment (salmans) [inline src/static/script.js:29]: Add a space after `=` 😛

**Review by salmans — APPROVED** [review]
LGTM

File: src/analyze/analysis_handler.go [hunk 8]
```diff
@@ -6,85 +6,148 @@ import (
...
+			_, resultVal := frequencies[timestamp]
+			if resultVal { //if there exists a value at a given time stamp in the final results
+				if frequencies[timestamp] < lengthMap[timestamp] { //check if the existing value is smaller than the incoming
+					frequencies[timestamp] = lengthMap[timestamp]
+				} else {
```
Comment (apai-fsly) [inline src/analyze/analysis_handler.go:51]: else is not needed here.

File: src/analyze/analysis_handler.go [hunk 9]
```diff
@@ -6,85 +6,148 @@ import (
...
+	//check the amount of the frequencies in results
+	if (len(AllResults.AllFreq) + 1) > maxPoints { //adding one more frequency would result in more than needed points
+		//remove the first item in the freq map; no modifications to the outlier map
+		firstResultKey := getSmallestTimestamp(getListOfKeys(frequencies))
+		delete(frequencies, firstResultKey)
+
+		//now update the results
+		for timestamp, _ := range lengthMap {
+			_, resultVal := frequencies[timestamp]
+			if resultVal { //if there exists a value at a given time stamp in the final results
```
Comment (apai-fsly) [inline src/analyze/analysis_handler.go:48]: you can make this all a single conditional or a switch statement. 

```
_, valPresent := frequencies[timestamp]
if valPresent && frequencies[timestamp] < lengthMap[timestamp] { dosomething...} 
else { dosomething else ....}

```

File: src/analyze/analysis_handler.go [hunk 10]
```diff
@@ -6,85 +6,148 @@ import (
...
-		MADTimestamps: bltOutlierTimes,
+		// Update outliers for both BLT MAD and ShakeAlert outliers
+		updateOutliers(AllResults, BLTMAD, bltIndexes, sortedTimestamps, sortedFrequencies)
+		updateOutliers(AllResults, SHAKEALERT, shakeAlertIndexes, sortedTimestamps, sortedFrequencies)
+
+	} else { //have not reached the max amount of points and can keep adding results
+		//modify the frequencies for the final results
+		for timestamp, _ := range lengthMap {
+			_, resultVal := frequencies[timestamp]
+			if resultVal { //if there exists a value at a given time stamp in the final results
```
Comment (apai-fsly) [inline src/analyze/analysis_handler.go:67]: same comment as above for condensing if statements.

File: src/analyze/analysis_handler.go [hunk 11]
```diff
@@ -6,85 +6,148 @@ import (
...
 
-//changed bltMad inputs to get timestamps and the outliers at the same time
-//this is also producing duplicates
-func BltMadWindow(data []float64, window common.Window, tau float64) ([]float64, []time.Time, [][]common.BGPMessage) {
-	var outliers []float64
-	var times []time.Time
-	var messages [][]common.BGPMessage //array of arrays of messages for a given bucket map
-	bucketMap := window.BucketMap
+//use the olderst timestamp when removing the data points
+func getSmallestTimestamp(timestamps []time.Time) time.Time {
```
Comment (apai-fsly) [inline src/analyze/analysis_handler.go:94]: make return type a pointer so you can return nil

File: src/analyze/analysis_handler.go [hunk 12]
```diff
@@ -6,85 +6,148 @@ import (
...
+
+		// If the outlier is not already in the map, create a new outlier for it
+		if !exists {
+			currResult.AllOutliers[outlierTimestamp] = createOutlierStruct(outlierTimestamp, algorithm, outlierFrequency)
+		} else { // If outlier not in map, need to possibly update it
+			if currOutlier.Count != outlierFrequency {
+				currOutlier.Count = outlierFrequency
+			}
+			// Update algorithm based on conditions
+			switch {
```
Comment (apai-fsly) [inline src/analyze/analysis_handler.go:125]: Lets avoid nesting if statements inside switch blocks.

File: src/analyze/analysis_handler.go [hunk 13]
```diff
@@ -6,85 +6,148 @@ import (
...
 }
+func createOutlierStruct(timestamp time.Time, algorithm int, count float64) common.OutlierInfo {
+	o := common.OutlierInfo{
+		Timestamp: timestamp,
+		Algorithm: algorithm,
+		Count:     count,
+	}
 
-//repeated code in three of these functions; moved outside to make the code easier to read
+	return o
```
Comment (apai-fsly) [inline src/analyze/analysis_handler.go:147]: you can make this one line - by

```
return common.OutlierInfo{field: value, field2:value2, etc...}
```

File: src/blt_mad/helpers.go [hunk 14]
```diff
@@ -20,7 +16,8 @@ func RemoveZeros(data []float64) ([]float64, error) {
 	}
```
Comment (sdstrowes) [inline src/blt_mad/helpers.go:17]: earlier in this file, `nonZeros` is counted by testing equality of a float (`if value != 0.0 {`) which is generally bad practice. And one way of handling this is more correctly is to test against a low threshold (e.g., `if math.Abs(value) < 0.00001`)

BUT: it made me wonder, why are these arrays of frequencies being passed around as floats? Are they not just integer counts?

File: src/main.go [hunk 15]
```diff
@@ -22,6 +22,7 @@ func dataHandler(w http.ResponseWriter, r *http.Request) {
 		http.Error(w, "Error encoding JSON", http.StatusInternalServerError)
 		return
 	}
+	//fmt.Println(analyze.AllResults)
```
Comment (sdstrowes) [inline src/main.go:25]: A useful thing is that git remembers everything. So commented sections of code can sometimes just be dropped; git has it, if you ever need it back!

File: src/analyze/analysis_handler.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (tyakovenko) [reply src/analyze/analysis_handler.go:40]: ^addressing in the filtering pr

File: src/analyze/analysis_handler.go (hunk 12 above)
Comment (tyakovenko) [reply src/analyze/analysis_handler.go:125]: ^addressing in the next pr

File: src/analyze/analysis_handler.go [hunk 16]
```diff
@@ -6,85 +6,148 @@ import (
...
+	for _, index := range indexes {
+		// Get outlier timestamp
+		outlierTimestamp := sortedTimestamps[index]
+		outlierFrequency := sortedFrequencies[index]
+
+		// Go through current outliers and see if timestamp exists
+		currOutlier, exists := currResult.AllOutliers[outlierTimestamp]
+
+		// If the outlier is not already in the map, create a new outlier for it
+		if !exists {
```
[the remark this answers is not in the document]
Comment (tyakovenko) [reply src/analyze/analysis_handler.go:118]: ^got refactored for filtering datastruct
